### PR TITLE
Format CSV output to allow for track-level audience reporting

### LINKED ISSUE #6: Populate audience measurement column using values from play counts generated by reporting script
In order to complete the SoundExchange reporting requirement, we need to fill in a comprehensive track list for each channel that lists how many times each track was played. The attached spreadsheet includes every track played on our channels, in show order. Filling out the audience measurement column seems like it would involve reading in the report generated by the script that totals how many times each show was played, matching on the show numbers, and then updating the audience measurement column with the numbers associated with each show.
[Show_track_master_list.xlsx](https://github.com/sul-dlss/rwj_reporting/files/213481/Show_track_master_list.xlsx)

### PR DESCRIPTION
**Pull request by jkeck**
Closes #6 

This PR starts w/ a refactor to make the `RWJReporter` class open to changing the CSV output, then swaps out the CSV Printer dependency in a separate commit.

### CODE REVIEW

File: lib/rwj_reporter.rb [hunk 1]
```diff
@@ -1,10 +1,8 @@
+require 'settings_file'
+require 'simple_csv_printer'
+require 'track_merged_csv_printer'
 
 class RwjReporter
-  def self.print_reports_for_dates(start_date_str, end_date_str, log_file_dir=nil, output_dir=nil)
-    rr = RwjReporter.new(start_date_str, end_date_str, log_file_dir)
-    rr.print_reports_for_dates(output_dir)
-  end
-
```
Comment (ndushay) [inline lib/rwj_reporter.rb:7]: I thought this premature optimization might be useful, like if they forgot to do a month or needed 6 months or something.  But I'm not emotionally attached to it.

**Comment by ndushay** [comment]
will data/track-template.csv   need updates over time?   Could there be a comment at the top explaining what it is?

**Comment by ndushay** [comment]
spec/lib/track_meged_...   has a typo in the filename

File: spec/lib/rwj_reporter_spec.rb [hunk 2]
```diff
@@ -102,19 +84,16 @@
...
       before(:context) do
-        RwjReporter.print_reports_for_dates('160228', '160303', FIXTURE_DIR, OUTPUT_DIR)
+        RwjReporter.new('160228', '160303', FIXTURE_DIR).print_reports_for_dates
       end
       let(:channel1_data_array) { File.open(File.join(OUTPUT_DIR, '160228-160303_channel_1_usage_counts.csv')).readlines }
       let(:channel2_data_array) { File.open(File.join(OUTPUT_DIR, '160228-160303_channel_2_usage_counts.csv')).readlines }
-      it 'each channel file is sorted by show number' do
-        expect(channel1_data_array).to eq channel1_data_array.sort
-        expect(channel2_data_array).to eq channel2_data_array.sort
-      end
```
Comment (ndushay) [inline spec/lib/rwj_reporter_spec.rb:112]: If we are still having channel files, should we still be ensuring they are sorted by show number?  Not clear to me why this test went away.

File: spec/lib/rwj_reporter_spec.rb [hunk 3]
```diff
@@ -55,28 +50,15 @@
       it 'takes output_dir from settings if no param' do
         # covered by other tests, as print_report_for_dates is called w/o arg
       end
-      it 'each channel file is sorted by show number' do
-        rr.print_reports_for_dates
-        expect(channel1_output_array).to eq channel1_output_array.sort
-        expect(channel2_output_array).to eq channel2_output_array.sort
-      end
```
Comment (ndushay) [inline spec/lib/rwj_reporter_spec.rb:62]: If we are still having channel files, should we still be ensuring they are sorted by show number? Not clear to me why this test went away.

**Comment by ndushay** [comment]
This seems ok to me.  I can't quite wrap my head around all the OO changes -- I've been too deep in other code, perhaps.   It's not clear to me all the refactoring is truly necessary, but it's good OO practice and good refactoring practice and it certainly isn't harming anything.

Does the README need to be updated?  Will the tracklist need updating? Will the user need to do anything differently?  Is the output different than stated?  Etc.

It would be great to have a "before" and "after" view of the output -- I am still not clear on exactly what was needed and exactly what is newly produced by these changes.   Maybe a little snippet of output expected in the class comments for the different printer implementations?

**Comment by jkeck** [comment]
@ndushay I believe I have addressed your comments with changes.

FWIW, I'll answer some of them here for posterity.
1. It has been explained that the master spreadsheet does not change.
2. Since we use the master spreadsheet as our template, we can expect that is sorted appropriately (now reflected in a class comment on the CSVPrinter class).
3. The new CSV output has been documented as a class comment on the CSVPrinter class.

## Example Output

### Old

This simply gave us the show number as the first column in the CSV, and the count as the second number.

```
6,1
7,2
8,3
```

### New

This output gives the show number as the first column, the track number as the 2nd column, and the count as the 3rd number (note that the count is intentionally repeated for each track number as we do not have track-level data)

```
6,1,0
6,2,0
6,3,0
7,1,2
7,2,2
7,3,2
7,4,2
8,1,3
8,2,3
8,3,3
```

**Comment by ndushay** [comment]
I feel better now.  👍  to merge.

**Comment by jmartin-sul** [comment]
i did not have a chance to thoroughly read through the whole thing, and did not have a chance to figure out whether i share one of @ndushay's concerns ("If we are still having channel files, should we still be ensuring they are sorted by show number? Not clear to me why this test went away.").

still, i did look through it as individual commits, and each of those seems reasonable and well tested to me.  so i am 👍, but hesitant to merge myself (since i can't spend more time on this before i get back wednesday).

**Comment by jkeck** [comment]
WRT @jmartin-sul's comment about the show numbers being sorted:

The sort order is already pre-determined by the track template.   Since the track template comes from the master spreadsheet (which the resulting CSV will be copied into), we **do not** want to do any sorting of the data as it would make the resulting CSV potentially not line up with the master spreadsheet from which the data originated.

**Pull request merged**